src/vmemcache.c

```diff
@@ -78,6 +237,149 @@ int
 vmemcache_put(VMEMcache *cache, const char *key, size_t ksize,
 				const char *value, size_t value_size)
 {
+	struct cache_entry *entry;
+	struct heap_entry he;
+
+	entry = Zalloc(sizeof(struct cache_entry) + ksize);
+	if (entry == NULL) {
+		ERR("!Zalloc");
+		return -1;
+	}
+
+	util_mutex_init(&entry->value.lock);
+	util_mutex_lock(&entry->value.lock);
+
+	entry->key.ksize = ksize;
+	memcpy(entry->key.key, key, ksize);
+
+	entry->value.p_entry =
+		cache->repl.ops->repl_p_insert(cache->repl.head, entry);
+
+	size_t left_to_allocate = value_size;
+
+	while (left_to_allocate > 0) {
+		he = vmcache_alloc(cache->heap, left_to_allocate);
+		if (HEAP_ENTRY_IS_NULL(he)) {
+			if (vmemcache_evict(cache, NULL, 0)) {
+				errno = ENOMEM;
+				goto error_no_mem;
+			}
+			continue;
+		}
+
+		if (VEC_PUSH_BACK(&entry->value.fragments, he))
+			FATAL("out of system memory");
+
+		if (left_to_allocate <= he.size)
+			break;
+
+		left_to_allocate -= he.size;
+	}
+
+	vmemcache_populate_fragments(entry, value, value_size);
+	entry->value.vsize = value_size;
+
+	ravl_insert(cache->index, entry);
+
+	util_mutex_unlock(&entry->value.lock);
+
+	return 0;
+
+error_no_mem:
+	VEC_FOREACH(he, &entry->value.fragments) {
+		vmcache_free(cache->heap, he);
+	}
+
+	VEC_DELETE(&entry->value.fragments);
+
+	util_mutex_unlock(&entry->value.lock);
+
+	return -1;
+}
+
+/*
+ * vmemcache_memcpy_fragments -- (internal) memcpy fragments
+ */
+static size_t
+vmemcache_memcpy_fragments(char *vbuf, size_t vbufsize, size_t offset,
```

Comment:
`vmemcache_populate_value`? to match the other function.

Reply:
Done